### PR TITLE
feat(acme): support redis namespace

### PR DESCRIPTION
**Pull request by catbro666**
### Summary

This PR relies on [fffonion/lua-resty-acme#101](https://github.com/fffonion/lua-resty-acme/pull/101)

`namespace` will be treated as a prefix of key and is default to empty string `""` for backward compatibility. `namespace` must not be prefixed with any of the reserved words.

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* Acme plugin now supports configuring `namespace` for redis storage which is default to empty string for backward compatibility.

### Issue reference

[KAG-615](https://konghq.atlassian.net/browse/KAG-615)

[KAG-615]: https://konghq.atlassian.net/browse/KAG-615?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kong/plugins/acme/schema.lua [hunk 1]
```diff
@@ -1,11 +1,24 @@
 local typedefs = require "kong.db.schema.typedefs"
+local reserved_words = require "kong.plugins.acme.reserved_words"
 
 local CERT_TYPES = { "rsa", "ecc" }
 
 local RSA_KEY_SIZES = { 2048, 3072, 4096 }
 
 local STORAGE_TYPES = { "kong", "shm", "redis", "consul", "vault" }
 
+local function validate_namespace(namespace)
```
Comment (fffonion) [inline kong/plugins/acme/schema.lua:10]: let's make sure to communicate with koko team on this constraint

File: kong-3.2.1-0.rockspec [hunk 2]
```diff
@@ -39,7 +39,7 @@ dependencies = {
   "lua-resty-openssl == 0.8.20",
   "lua-resty-counter == 0.2.1",
   "lua-resty-ipmatcher == 0.6.1",
-  "lua-resty-acme == 0.10.1",
+  "lua-resty-acme == 0.11.0",
```
[the remark this answers is not in the document]
Comment (catbro666) [reply kong-3.2.1-0.rockspec:42]: done.

File: spec/03-plugins/29-acme/05-redis_storage_spec.lua [hunk 3]
```diff
@@ -22,4 +23,155 @@ describe("Plugin: acme (storage.redis)", function()
...
+
+    local err = storage2:set("foo", "2", 10)
+    assert.is_nil(err)
+    local value, err = storage0:get("foo")
+    assert.is_nil(err)
+    assert.equal("0", value)
+    local value, err = storage1:get("foo")
+    assert.is_nil(err)
+    assert.equal("1", value)
+    local value, err = storage2:get("foo")
```
[the remark this answers is not in the document]
Comment (catbro666) [reply spec/03-plugins/29-acme/05-redis_storage_spec.lua]: done.

File: kong/plugins/acme/schema.lua (hunk 1 above)
Comment (catbro666) [reply kong/plugins/acme/schema.lua:10]: OK, will inform the koko team on the JIRA ticket.

File: CHANGELOG.md [hunk 4]
```diff
@@ -104,6 +104,8 @@
   [#10533](https://github.com/Kong/kong/pull/10533)
 - **Zipkin&Opentelemetry**: convert traceid in http response headers to hex format
   [#10534](https://github.com/Kong/kong/pull/10534)
+- **ACME**: acme plugin now supports configuring `namespace` for redis storage
+  which is default to empty string for backward compatibility.
```
Comment (chronolaw) [inline CHANGELOG.md:108]:
```suggestion
  which is default to empty string for backward compatibility.
  [#10562](https://github.com/Kong/kong/pull/10562)
```

File: kong/plugins/acme/schema.lua [hunk 5]
```diff
@@ -1,11 +1,24 @@
...
+  if namespace ~= "" then
+    for _, v in pairs(reserved_words) do
+      if namespace:sub(1, #v) == v then
```
Comment (chronolaw) [inline kong/plugins/acme/schema.lua:13]: should we use `pl.startswith(s, prefix)`?

Comment (chronolaw) [reply kong/plugins/acme/schema.lua:13]: Or `namespace:find(v, 1, true) == 1`?

**Review by fffonion — APPROVED** [review]

File: kong/plugins/acme/schema.lua (hunk 5 above)
Comment (catbro666) [reply kong/plugins/acme/schema.lua:13]: Not sure which one is more efficient. `string.sub` generates a new string while `string.find` may need to traverse the whole string.

Comment (chronolaw) [reply kong/plugins/acme/schema.lua:13]: OK, let's keep this line.

**Review by fffonion — APPROVED** [review]
merge on green

**Comment by hbagdi** [comment]
@fffonion @hanshuebner @catbro666 Why was this PR merged without a corresponding documentation PR?

**Comment by hbagdi** [comment]
@catbro666 Please open the corresponding EE PR.

**Comment by catbro666** [comment]
Opened the corresponding EE/DOC PR.

**Comment by hbagdi** [comment]
@hanshuebner @catbro666 Why was this commit merged without any tests that assert that the namespace parameter is used and takes effect or not?

**Comment by catbro666** [comment]
@hbagdi Because the storage_config is passed directly to the bottom library. There's nothing special about the namespace parameter compared to the other parameters. So I think the current tests can ensure that it takes effect.

**Comment by hbagdi** [comment]
Is there a test that proves that the configuration parameter is correctly passed down to the library?

**Comment by hbagdi** [comment]
ping @catbro666

**Comment by catbro666** [comment]
@hbagdi Sorry, I may have misunderstood you before. 
> any tests that assert that the namespace parameter is used and takes effect or not?
I think [this test](https://github.com/Kong/kong/blob/990e539905de7da72f8071f4368a4ef7a055f6a1/spec/03-plugins/29-acme/05-redis_storage_spec.lua#L109) can cover this. It creates 3 namespaces, one default and two specified. They can work independently and don't interfere with each other.

**Comment by hbagdi** [comment]
That test tests that the library can take namespace as an input and that there is isolation between namespaces.
It doesn't prove that if a user configures a 'namespace' in the acme plugin configuration, the code in this repository will indeed configure the library accordingly.
In other words, there is no test that would fail if the namespace configuration input is ignored by Kong and not passed into the underlying library.

Does that make sense?

**Comment by catbro666** [comment]
> It doesn't prove that if a user configures a 'namespace' in the acme plugin configuration, the code in this repository will indeed configure the library accordingly.

Indeed, that test can't prove this. But this part is a common logic shared by all storage and independent of the specific storage implementation. The storage configuration is transparent to the plugin, the plugin just simply passes it down the library. As for proving the storage_config can be successfully passed to the library, we already have corresponding tests in [01-client_spec.lua](https://github.com/Kong/kong/blob/990e539905de7da72f8071f4368a4ef7a055f6a1/spec/03-plugins/29-acme/01-client_spec.lua) and [03_access_spec.lua](https://github.com/Kong/kong/blob/990e539905de7da72f8071f4368a4ef7a055f6a1/spec/03-plugins/29-acme/03-access_spec.lua).

**Comment by hbagdi** [comment]
> As for proving the storage_config can be successfully passed to the library, we already have corresponding tests in [01-client_spec.lua](https://github.com/Kong/kong/blob/990e539905de7da72f8071f4368a4ef7a055f6a1/spec/03-plugins/29-acme/01-client_spec.lua) and [03_access_spec.lua](https://github.com/Kong/kong/blob/990e539905de7da72f8071f4368a4ef7a055f6a1/spec/03-plugins/29-acme/03-access_spec.lua).

I'm struggling so please bear with me.
I don't see a unit test that ensures that the configuration from the schema is translated into the configuration table passed down to the underlying library correctly.

**Comment by catbro666** [comment]
@hbagdi Added tests in this [PR](https://github.com/Kong/kong/pull/10632) to ensure that the namespace parameter is used and takes effect.

**Pull request merged**